### DISCUSSION TITLE
Minver does not effect Version and ApplicationDisplayVersion of a .Net MAUI project

### DISCUSSION THREAD

**Discussion by Peaj** [adamralph/minver#938] [General]
I am trying to version a .Net MAUI project.
Minver does correctly determine the version from git as seen in the following (detailed) log:

```
MinVer: [input] MSBuildProjectDirectory=D:\a\MetricTimer\MetricTimer\MetricTimer
  MinVer: [input] MinVerAutoIncrement=
  MinVer: [input] MinVerBuildMetadata=
  MinVer: [input] MinVerDefaultPreReleaseIdentifiers=
  MinVer: [input] MinVerDefaultPreReleasePhase=
  MinVer: [input] MinVerIgnoreHeight=
  MinVer: [input] MinVerMinimumMajorMinor=
  MinVer: [input] MinVerTagPrefix=v
  MinVer: [input] MinVerVerbosity=detailed
  MinVer: [input] MinVerVersionOverride=
  MinVer: MinVer 4.3.0+c3754a551d7c9e17d63798dd8d33d1e309e464f9.
  MinVer: Ignoring non-version tag { Name: v0.1, Sha: ffef46bb7f75ef242fcb5a33bf4ad3a5a18aa231 }.
  MinVer: Ignoring non-version tag { Name: v0.1, Sha: 8414563ecdec44d656e65077560abcc53768fbc4 }.
  MinVer: Ignoring non-version tag { Name: v0.2, Sha: 082dabf23cda9041f8f85b2dbb386914c50b75cd }.
  MinVer: Ignoring non-version tag { Name: v0.3, Sha: 9f187098d76b979ee56b97facbf4cc777ed3fd08 }.
  MinVer: Ignoring non-version tag { Name: v0.4, Sha: 4329a973292b32cd77c38d825ec3bc25101704ea }.
  MinVer: Ignoring non-version tag { Name: v0.6, Sha: dafafbbad4c77b81135fe6d0bfe90dd233ea331f }.
  MinVer: Ignoring non-version tag { Name: v0.7, Sha: 8a56df0a6411dd214354197e58b05c8ff1d02816 }.
  MinVer: 3 commits checked.
  MinVer: Using { Commit: 4cfaff7, Tag: 'v0.7.2', Version: 0.7.2, Height: 2 }.
  MinVer: The calculated version 0.7.3-alpha.0.2 satisfies the minimum major minor 0.0.
  MinVer: Calculated version 0.7.3-alpha.0.2.
  MinVer: [output] MinVerVersion=0.7.3-alpha.0.2
  MinVer: [output] MinVerMajor=0
  MinVer: [output] MinVerMinor=7
  MinVer: [output] MinVerPatch=3
  MinVer: [output] MinVerPreRelease=alpha.0.2
  MinVer: [output] MinVerBuildMetadata=
  MinVer: [output] AssemblyVersion=0.0.0.0
  MinVer: [output] FileVersion=0.7.3.0
  MinVer: [output] PackageVersion=0.7.3-alpha.0.2
  MinVer: [output] Version=0.7.3-alpha.0.2
  ```
It does however not seem to have any effect on the various versions of the build.
When installing the package and also when getting the version in code via `AppInfo.Current.VersionString` it still displays the versions originally set in the csproj file via `Version` and/or `ApplicationDisplayVersion`.
I am unsure where to find/get the Assembly version in a MAUI built to validate it.

I also tried to explicitly set the versions as detailed in the custom section but it does not have any effect either:

```XML
<Target Name="UpdateAssemblyVersion" AfterTargets="MinVer">
  <PropertyGroup>
    <AssemblyVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch).0</AssemblyVersion>
    <Version>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)-$(MinVerPreRelease)</Version>
    <ApplicationDisplayVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)-$(MinVerPreRelease)</ApplicationDisplayVersion>
  </PropertyGroup>
</Target>
```

**Comment by Peaj** [comment]
I added messages to output the generated versions to the console:

```XML
<Target Name="SpecialTargetTestStep" BeforeTargets="BeforeBuild">
  <Message Text="TEST STEP" Importance="high"/>
</Target>
<Target Name="UpdateAssemblyVersion" AfterTargets="MinVer">
  <PropertyGroup>
    <AssemblyVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch).0</AssemblyVersion>
    <Version>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)-$(MinVerPreRelease)</Version>
    <PackageVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)-$(MinVerPreRelease)</PackageVersion>
    <ApplicationDisplayVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)-$(MinVerPreRelease)</ApplicationDisplayVersion>
  </PropertyGroup>
  <Message Text="Update Assembly Version" Importance="high"/>
  <Message Text="AssemblyVersion = $(AssemblyVersion)" Importance="high" />
  <Message Text="Version = $(Version)" Importance="high" />
  <Message Text="PackageVersion = $(PackageVersion)" Importance="high" />
  <Message Text="ApplicationDisplayVersion = $(ApplicationDisplayVersion)" Importance="high" />
</Target>
```
And it seems to be correct:
```
1>Update Assembly Version
1>AssemblyVersion = 0.7.3.0
1>Version = 0.7.3-alpha.0.1
1>PackageVersion = 0.7.3-alpha.0.1
1>ApplicationDisplayVersion = 0.7.3-alpha.0.1
```
Maybe MinVer is running to late in the build process?
I guess MAUI might be generating app manifest files from the csproj data before MinVer can run.

**Reply by Peaj** [reply]
I debugged the MSBuild process and my suspicion seems to be right.
The `MauiGeneratePackageAppxManifest` Target runs before the `MinVer` target when the Versions have not been written yet.
I guess the solution would be to run the `MinVer` step earlier in the build.

**Reply by adamralph** [reply]
@Peaj well done! I always find it difficult to figure these things out. I've never even looked at a MAUI project so I have no reference point, but I'd be happy to accept a pull request which fixes this for you, as long as it doesn't break anything for other project types.

**Reply by Peaj** [reply]
Well thank you for that great tool. I am pretty new to MSBuild but fortunately I found that nice blog entry regarding MSBuild debugging: https://dfederm.com/debugging-msbuild/
I have a workaround for now which I will post as an answer to this question.
Right now I am running `MinVer` before `MauiGeneratePackageAppxManifest` which is MAUI specific.

I will see what I can do regarding a general solution for a PR.
MAUI version is quite tricky as it has at least 4 different kind of versions with different schemes (many not SemVer compatible).
So getting it to work (and test) on all platforms and keeping it compatible with non MAUI projects is the tricky part.

Does MinVer have any dependencies it has to run after or can it just be run as early as possible?
This info might help me to find the proper position to run it at for a general solution.

**Reply by Peaj** [reply]
Nerdbank.Gitversioning did seem to have a pretty similar problem with MAUI in the past:
https://github.com/dotnet/Nerdbank.GitVersioning/issues/828

**Comment by Peaj** [comment]
The problem is that MinVar runs after the platform specific Manifest files have been generated so the version does not effect the actual packages.

This is a workaround making the MinVar build step run earlier to update the version numbers before app manifests are generated:

```XML
<!-- Workaround to make MinVer run earlier -->
<!-- This allows to update Versions before platform manifest files are generated -->
<!-- It also sets ApplicationDisplayVersion and AssemblyVersion -->
<Target Name="MinVerPrerun" BeforeTargets="MauiGeneratePackageAppxManifest">
  <Message Text="MinVer Prerun" Importance="high"/>
  <MSBuild Projects="$(MSBuildProjectFile)" Targets="MinVer" />
</Target>
<Target Name="UpdateVersions" AfterTargets="MinVerPrerun">
  <PropertyGroup>
    <AssemblyVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch).0</AssemblyVersion>
    <ApplicationDisplayVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)</ApplicationDisplayVersion>
  </PropertyGroup>
  <Message Text="Update Versions" Importance="high"/>
  <Message Text="AssemblyVersion = $(AssemblyVersion)" Importance="high" />
  <Message Text="Version = $(Version)" Importance="high" />
  <Message Text="PackageVersion = $(PackageVersion)" Importance="high" />
  <Message Text="ApplicationDisplayVersion = $(ApplicationDisplayVersion)" Importance="high" />
</Target>
```

As ApplicationDisplayVersion is not set by MinVar and AssemblyVersion is set to 0.0.0.0 it is updated manually in the second step.
ApplicationDisplayVersion uses a simplified version as it only allows 3 separated numbers.
This has only been tested on windows builds for now.

**Reply by adamralph** [reply]
@Peaj thanks for posting your workaround.

Would this also work?

```xml
<Target Name="MinVerMaui" BeforeTargets="MauiGeneratePackageAppxManifest" DependsOnTargets="MinVer">
  <PropertyGroup>
    <ApplicationDisplayVersion>$(MinVerMajor).$(MinVerMinor).$(MinVerPatch)</ApplicationDisplayVersion>
  </PropertyGroup>
  <Message Importance="$(MinVerDetailed)" Text="ApplicationDisplayVersion = $(ApplicationDisplayVersion)" />
</Target>
```

It seems like `ApplicationDisplayVersion` is the only additional thing that needs to be set for your MAUI project.

MinVer [already sets `AssemblyVersion`](https://github.com/adamralph/minver#version-numbers) and setting it to something else seems like a separate concern, not related to MAUI (the FAQ already contains an [example showing how to set it to something else](https://github.com/adamralph/minver#version-numbers)). Or does MAUI somehow require `AssemblyVersion` to be `$(MinVerMajor).$(MinVerMinor).$(MinVerPatch).0`?

**Reply by Peaj** [reply]
Thats a great idea to fix the dependencies this way. This does actually work. Only had to fix the typo with DependsOnTargets.
I just assumed the Assembly version has not been set at all because it was 0.0.0.0 but now I understand that it is intended and is just setting the major version (which was also 0 in my case)
So I guess the only thing left is making this step conditional to run only in MAUI projects and to check if it works on different platforms then windows.

**Reply by Peaj** [reply]
I am unsure if this will produce version collisions though. As the Version and Displayversion are missing the pre release tags it may result in multiple versions with the same version number. 0.7.3-alpha.0.1 will be the same as 0.7.3.1.
In MAUI there is also the ApplicationVersion which is just a single number that has to increase with each built.
I think it is kind of like a built number. Do you have any recommendation what to set this to?
The version number you can get at runtime seems to be constructed from `ApplicationDisplayVersion`.`ApplicationVersion`

**Reply by bording** [reply]
> I am unsure if this will produce version collisions though. As the Version and Displayversion are missing the pre release tags it may result in multiple versions with the same version number. 0.7.3-alpha.0.1 will be the same as 0.7.3.1.

If `ApplicationDisplayVersion` supports full SemVer version strings, then you can just change what you are what value you are setting it to. You can use `MinVerVersion` to get the full value, for example.

**Reply by Peaj** [reply]
The problem is `ApplicationDisplayVersion` does only allow 3 dot separated numbers.
Now that I had time to try it out this obviously leads to version conflicts as the height will not effect the build number.
In case of MAUI app packages (on windows) this means I cannot update to those versions as it says the current version is higher.

The best solution I can come up with would be to set the `ApplicationVersion` (Which is essentially a build number) to the `Height`
The `ApplicationVersion` is added as the 4th component to the build number by MAUI.
So this will at least result in different versions like `ApplicationDisplayVersion` . `Height`

In the case from the original post this would transform 0.7.3-alpha.0.2 into 0.7.3.2.
Is there any variable that just contains the height without the "alpha.0." string as I could not find it in the docs?

**Reply by adamralph** [reply]
> Is there any variable that just contains the height without the "alpha.0." string as I could not find it in the docs?

No.

BTW, you shouldn't really be using any version with height for releases. The idea is that, while those versions can be distinguished, when you actually want to release (including pre-release) you tag the repo. And in your tags you can increment the through `alpha.1`, `alpha.2`, `beta.1`, `beta.2`, etc. etc. If MAUIs versioning scheme is restricted to three numerics only, then it sounds like it doesn't really cater for pre-releases at all. I'm not sure what the solution is here, but I don't think using height is it.

**Reply by adamralph** [reply]
related — https://github.com/adamralph/minver?tab=readme-ov-file#what-is-height-for
